### PR TITLE
Add checks for MAVLink RangeFinder's min/max fields

### PR DESCRIPTION
**Pull request by peterbarker**
We consume `DISTANCE_SENSOR` messages in ArduPilot for a "mavlink rangefinder".

At the moment we ignore the `min_distance` and `max_distance` fields in the message.

This PR changes that.

More autotests are added to make sure we honour those parameters.

This replaces https://github.com/ArduPilot/ardupilot/pull/9338/files by @gs-niteesh - @OXINARF had asked for something similar to this (state doesn't hold min/max, so they're held as attributes in the mavlink subclass instead in this PR).

### CODE REVIEW

**Review by khancyr — APPROVED** [review]
LGTM

**Comment by OXINARF** [comment]
> This replaces https://github.com/ArduPilot/ardupilot/pull/9338/files by @gs-niteesh - @OXINARF had asked for something similar to this (state doesn't hold min/max, so they're held as attributes in the mavlink subclass instead in this PR).

@peterbarker Back then the state held the parameters. We still have the min/max distance parameters and I see no reason why they shouldn't be reused here instead of adding new variables that replace the parameters.

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -4294,6 +4294,62 @@ def wait_gps_sys_status_not_present_or_enabled_and_healthy(self, timeout=30):
...
+                return False
+        return True
+
+    def wait_prearm_sys_status_healthy(self, timeout=60):
+        tstart = self.get_sim_time_cached()
+        while True:
+            t2 = self.get_sim_time_cached()
+            if t2 - tstart > timeout:
+                self.progress("Prearm bit never went true")
+                self.arm_vehicle()
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/autotest/common.py:4348]: I added this comment:
```
                self.progress("Prearm bit never went true.  Attempting arm to elicit reason from autopilot")

```

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 2]
```diff
@@ -28,8 +28,8 @@ void AP_RangeFinder_MAVLink::handle_msg(const mavlink_message_t &msg)
     if (packet.orientation == MAV_SENSOR_ROTATION_PITCH_270) {
         state.last_reading_ms = AP_HAL::millis();
         distance_cm = packet.current_distance;
+        sensor_type = (MAV_DISTANCE_SENSOR)packet.type;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:33]: Interesting thought.  We could use the parameter on the rangefinder indicating what direction it is and only accept measurements from packets with that orientation.

We're also a little indiscriminate about who we accept these messages from.  If you're on a mesh network wackiness will ensue.

**Comment by peterbarker** [comment]
> > This replaces https://github.com/ArduPilot/ardupilot/pull/9338/files by @gs-niteesh - @OXINARF had asked for something similar to this (state doesn't hold min/max, so they're held as attributes in the mavlink subclass instead in this PR).
> 
> @peterbarker Back then the state held the parameters. We still have the min/max distance parameters and I see no reason why they shouldn't be reused here instead of adding new variables that replace the parameters.

We've learn that the code setting its own values into those variables can be very painful - we've had several cases where users have downloaded those values to their GCS and restored them later - thus values which were supposed to be temporary become permanent.  We're removing many cases where we've done this so adding one sounds like a bad idea :-)

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 3]
```diff
@@ -55,8 +28,10 @@ void AP_RangeFinder_MAVLink::handle_msg(const mavlink_message_t &msg)
     if (packet.orientation == MAV_SENSOR_ROTATION_PITCH_270) {
         state.last_reading_ms = AP_HAL::millis();
         distance_cm = packet.current_distance;
+        _max_distance_cm = packet.max_distance;
+        _min_distance_cm = packet.min_distance;
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:32]: should use larger of the minimums, and smaller of the maximums
also, do we need a special case for magic values like 0 or UINT16_MAX?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:32]: Done.

No magic values in this message.

**Review by tridge — APPROVED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp [hunk 4]
```diff
@@ -55,8 +28,25 @@ void AP_RangeFinder_MAVLink::handle_msg(const mavlink_message_t &msg)
...
+        sensor_type = (MAV_DISTANCE_SENSOR)packet.type;
+    }
+}
+
+int16_t AP_RangeFinder_MAVLink::max_distance_cm() const
+{
+    if (params.max_distance_cm < _max_distance_cm) {
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp:39]: if min and max distance in pkt are both zero then ignore both of them

**Review by tridge — CHANGES_REQUESTED** [review]

**Pull request merged**